File: neurmq_ffmpeg.py
import zmq
import time
import asyncio
import cv2
import numpy as np
import pyaudio
import ffmpeg
import logging

async def main():
    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)

    context = zmq.Context()
    subscriber = context.socket(zmq.SUB)
    subscriber.connect("tcp://localhost:5555")
    topics = ["SET_JOINT_FORCES", "GET_DOF_START_INDICES", "GET_JOINT_POSITIONS",
              "GET_JOINT_VELOCITIES", "GET_JOINT_FORCES", "GET_JOINT_EXTERNAL_FORCES", "ERROR"]
    for topic in topics:
        subscriber.subscribe(topic)
    logger.info("Subscriber connected to tcp://localhost:5555")

    publisher = context.socket(zmq.PUB)
    publisher.connect("tcp://localhost:5556")
    logger.info("Publisher connected to tcp://localhost:5556")

    time.sleep(1)

    base_port = 8888
    audio_port = 8887
    max_cameras = 10
    stream_processes = []
    stream_dimensions = []
    audio_output = None
    pyaudio_instance = None

    try:
        # Capture audio stream
        audio_url = f"udp://127.0.0.1:{audio_port}"
        try:
            probe = ffmpeg.probe(audio_url, timeout=2)
            audio_stream = next((s for s in probe['streams'] if s['codec_type'] == 'audio'), None)
            if audio_stream:
                channels = int(audio_stream['channels'])
                sample_rate = int(audio_stream['sample_rate'])
                pyaudio_instance = pyaudio.PyAudio()
                audio_output = pyaudio_instance.open(
                    format=pyaudio.paInt16,
                    channels=channels,
                    rate=sample_rate,
                    output=True
                )
                audio_process = (
                    ffmpeg
                    .input(audio_url, f='webm')
                    .output('pipe:', format='s16le', ac=channels, ar=sample_rate)
                    .run_async(pipe_stdout=True, pipe_stderr=True)
                )
                stream_processes.append(audio_process)
                logger.info(f"Started capturing audio at {audio_url} with {channels} channels")
        except ffmpeg.Error as e:
            logger.warning(f"No audio stream at {audio_url}: {e.stderr.decode()}")

        # Capture video streams
        for port in range(base_port, base_port + max_cameras):
            stream_url = f"udp://127.0.0.1:{port}"
            try:
                probe = ffmpeg.probe(stream_url, timeout=2)
                video_stream = next((s for s in probe['streams'] if s['codec_type'] == 'video'), None)
                if not video_stream:
                    logger.info(f"No video stream at {stream_url}")
                    continue

                width = int(video_stream['width'])
                height = int(video_stream['height'])
                stream_dimensions.append((width, height, port))

                process = (
                    ffmpeg
                    .input(stream_url, f='webm')
                    .output('pipe:', format='rawvideo', pix_fmt='bgr24')
                    .run_async(pipe_stdout=True, pipe_stderr=True)
                )
                stream_processes.append(process)
                logger.info(f"Started capturing video at {stream_url}")
            except ffmpeg.Error as e:
                logger.info(f"No stream at {stream_url}: {e.stderr.decode()}")
                break

        if not stream_processes and not audio_output:
            logger.error("No streams detected. Exiting.")
            return

        try:
            while True:
                # Read video frames
                for i, (process, (width, height, port)) in enumerate(zip(stream_processes, stream_dimensions)):
                    if 'rawvideo' in process.args:  # Video process
                        in_bytes = process.stdout.read(width * height * 3)
                        if not in_bytes:
                            logger.warning(f"End of video stream at port {port}")
                            break
                        frame = np.frombuffer(in_bytes, np.uint8).reshape(height, width, 3)
                        cv2.imshow(f"Unity Stream {port}", frame)

                # Read audio
                if audio_output:
                    audio_bytes = stream_processes[0].stdout.read(4096)
                    if audio_bytes:
                        audio_output.write(audio_bytes)

                # Handle NetMQ
                try:
                    topic = subscriber.recv_string(zmq.NOBLOCK)
                    message = subscriber.recv_string()
                    logger.info(f"Received: {topic}:{message}")
                except zmq.Again:
                    pass

                print("\nAvailable commands:")
                print("1. Set joint forces (e.g., 100,200,300)")
                print("2. Get DOF start indices")
                print("3. Get joint positions")
                print("4. Get joint velocities")
                print("5. Get joint forces")
                print("6. Get joint external forces")
                print("7. Exit")
                choice = input("Enter choice (1-7): ")

                command = None
                if choice == "1":
                    forces = input("Enter forces (comma-separated, e.g., 100,200,300): ")
                    command = f"SET_JOINT_FORCES:{forces}"
                elif choice == "2":
                    command = "GET_DOF_START_INDICES:"
                elif choice == "3":
                    command = "GET_JOINT_POSITIONS:"
                elif choice == "4":
                    command = "GET_JOINT_VELOCITIES:"
                elif choice == "5":
                    command = "GET_JOINT_FORCES:"
                elif choice == "6":
                    command = "GET_JOINT_EXTERNAL_FORCES:"
                elif choice == "7":
                    print("Exiting...")
                    break
                else:
                    print("Invalid choice, please try again.")
                    continue

                if command:
                    publisher.send_string(command)
                    logger.info(f"Sent: {command}")

                if cv2.waitKey(1) & 0xFF == ord('q'):
                    break

        finally:
            for process in stream_processes:
                process.terminate()
            if audio_output:
                audio_output.stop_stream()
                audio_output.close()
            if pyaudio_instance:
                pyaudio_instance.terminate()
            cv2.destroyAllWindows()

    except Exception as e:
        logger.error(f"Error capturing streams: {e}")

    finally:
        subscriber.close()
        publisher.close()
        context.term()

if __name__ == "__main__":
    asyncio.run(main())

File: neurmq_ffmpeg.cs
using UnityEngine;
using NetMQ;
using NetMQ.Sockets;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FFMpegCore;
using FFMpegCore.Pipes;
using System.Threading.Tasks;
using System.Diagnostics;

[RequireComponent(typeof(ArticulationBody), typeof(AudioListener))]
public class ArticulationBodyNetMQ : MonoBehaviour
{
    [SerializeField] private string publisherAddress = "tcp://*:5555";
    [SerializeField] private string subscriberAddress = "tcp://*:5556";
    [SerializeField] private int baseUdpPort = 8888;
    [SerializeField] private int audioUdpPort = 8887;
    [SerializeField, Min(1)] private int renderWidth = 1280;
    [SerializeField, Min(1)] private int renderHeight = 720;

    private ArticulationBody articulationBody;
    private PublisherSocket publisherSocket;
    private SubscriberSocket subscriberSocket;
    private List<CameraStream> cameraStreams = new List<CameraStream>();
    private List<float> audioBuffer = new List<float>();
    private bool isStreaming = false;
    private List<Task> ffmpegTasks = new List<Task>();
    private int audioChannels;

    private class CameraStream
    {
        public Camera Camera;
        public RenderTexture RenderTexture;
        public Texture2D TempTexture;
        public int UdpPort;
    }

    void Reset()
    {
        articulationBody = GetComponent<ArticulationBody>();
        if (articulationBody == null)
        {
            Debug.LogError("ArticulationBody component not found.", this);
            return;
        }

        Camera[] cameras = GetComponentsInChildren<Camera>();
        if (cameras.Length == 0)
        {
            Debug.LogWarning("No Cameras found in hierarchy. Video streaming unavailable.", this);
        }

        cameraStreams.Clear();
        foreach (var camera in cameras)
        {
            var stream = new CameraStream
            {
                Camera = camera,
                UdpPort = baseUdpPort + cameraStreams.Count
            };
            UpdateCameraTextures(stream);
            cameraStreams.Add(stream);
            Debug.Log($"Camera {camera.name} assigned, will stream to udp://127.0.0.1:{stream.UdpPort}");
        }
    }

    void OnValidate()
    {
        if (renderWidth < 1) renderWidth = 1;
        if (renderHeight < 1) renderHeight = 1;

        foreach (var stream in cameraStreams)
        {
            UpdateCameraTextures(stream);
        }
    }

    void UpdateCameraTextures(CameraStream stream)
    {
        if (stream.RenderTexture != null && (stream.RenderTexture.width != renderWidth || stream.RenderTexture.height != renderHeight))
        {
            Destroy(stream.RenderTexture);
            stream.RenderTexture = null;
        }
        if (stream.TempTexture != null && (stream.TempTexture.width != renderWidth || stream.TempTexture.height != renderHeight))
        {
            Destroy(stream.TempTexture);
            stream.TempTexture = null;
        }

        if (stream.RenderTexture == null)
        {
            stream.RenderTexture = new RenderTexture(renderWidth, renderHeight, 24);
            stream.Camera.targetTexture = stream.RenderTexture;
        }
        if (stream.TempTexture == null)
        {
            stream.TempTexture = new Texture2D(renderWidth, renderHeight, TextureFormat.RGB24, false);
        }
    }

    void Start()
    {
        articulationBody = GetComponent<ArticulationBody>();
        if (articulationBody == null)
        {
            Debug.LogError("ArticulationBody missing.", this);
            return;
        }

        // Check FFmpeg binaries
        string ffmpegLocalPath = Path.Combine(Application.dataPath, "FFmpeg", "bin", "ffmpeg");
        string ffmpegPath = ffmpegLocalPath;
        bool usingSystemFfmpeg = false;

        if (!File.Exists(ffmpegLocalPath))
        {
            try
            {
                Process process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "ffmpeg",
                        Arguments = "-version",
                        RedirectStandardOutput = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };
                process.Start();
                string output = process.StandardOutput.ReadToEnd();
                process.WaitForExit();
                if (process.ExitCode == 0)
                {
                    Debug.LogWarning($"FFmpeg not found at {ffmpegLocalPath}, but system FFmpeg detected. Using system FFmpeg.");
                    ffmpegPath = "ffmpeg";
                    usingSystemFfmpeg = true;
                }
                else
                {
                    Debug.LogError($"FFmpeg not found at {ffmpegLocalPath} or in system PATH. Please place ffmpeg in Assets/FFmpeg/bin or ensure ffmpeg is in PATH.", this);
                    return;
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"FFmpeg not found at {ffmpegLocalPath} or in system PATH. Error checking system FFmpeg: {e.Message}. Place ffmpeg in Assets/FFmpeg/bin or add to PATH.", this);
                return;
            }
        }
        else
        {
            Debug.Log($"Using FFmpeg at {ffmpegLocalPath}");
        }

        if (!usingSystemFfmpeg)
        {
            GlobalFFOptions.Configure(new FFOptions { BinaryFolder = Path.Combine(Application.dataPath, "FFmpeg", "bin") });
        }

        // Determine audio channels
        audioChannels = AudioSettings.speakerMode switch
        {
            AudioSpeakerMode.Mono => 1,
            AudioSpeakerMode.Stereo => 2,
            AudioSpeakerMode.Quad => 4,
            AudioSpeakerMode.Surround => 5,
            AudioSpeakerMode.Mode5point1 => 6,
            AudioSpeakerMode.Mode7point1 => 8,
            _ => 2 // Default to stereo
        };
        Debug.Log($"Detected {audioChannels} audio channels based on AudioSettings.speakerMode.");

        // Initialize NetMQ
        AsyncIO.ForceDotNet.Force();
        publisherSocket = new PublisherSocket();
        try
        {
            publisherSocket.Bind(publisherAddress);
            Debug.Log($"Publisher socket bound to {publisherAddress}");
        }
        catch (Exception e)
        {
            Debug.LogError($"Failed to bind publisher socket: {e.Message}");
            return;
        }

        subscriberSocket = new SubscriberSocket();
        try
        {
            subscriberSocket.Bind(subscriberAddress);
            subscriberSocket.Subscribe("");
            Debug.Log($"Subscriber socket bound to {subscriberAddress}");
        }
        catch (Exception e)
        {
            Debug.LogError($"Failed to bind subscriber socket: {e.Message}");
            publisherSocket?.Close();
            publisherSocket?.Dispose();
            return;
        }

        // Start FFmpeg streaming
        isStreaming = true;
        ffmpegTasks.Add(Task.Run(() => StreamAudio()));
        foreach (var stream in cameraStreams)
        {
            ffmpegTasks.Add(Task.Run(() => StreamCamera(stream)));
        }
        Debug.Log($"Started FFmpeg streaming: {cameraStreams.Count} cameras, audio to udp://127.0.0.1:{audioUdpPort}");
    }

    void OnAudioFilterRead(float[] data, int channels)
    {
        if (!isStreaming) return;
        lock (audioBuffer)
        {
            audioBuffer.AddRange(data);
        }
    }

    async void StreamCamera(CameraStream stream)
    {
        try
        {
            var videoSource = new RawVideoPipeSource(GetVideoFrames(stream))
            {
                FrameRate = 30,
                Width = renderWidth,
                Height = renderHeight,
                PixelFormat = "rgb24"
            };

            await FFMpegArguments
                .FromPipeInput(videoSource, options => options.WithCustomArgument("-re"))
                .OutputToUrl($"udp://127.0.0.1:{stream.UdpPort}", options => options
                    .WithVideoCodec("libvpx-vp9")
                    .ForceFormat("webm")
                    .WithCustomArgument("-b:v 2M"))
                .ProcessAsynchronously();
        }
        catch (Exception ex)
        {
            Debug.LogError($"FFmpeg streaming error for camera {stream.Camera.name}: {ex.Message}");
        }
    }

    async void StreamAudio()
    {
        try
        {
            var audioSource = new RawAudioPipeSource(GetAudioSamples())
            {
                SampleRate = AudioSettings.outputSampleRate,
                Channels = audioChannels,
                SampleFormat = "s16"
            };

            await FFMpegArguments
                .FromPipeInput(audioSource)
                .OutputToUrl($"udp://127.0.0.1:{audioUdpPort}", options => options
                    .WithAudioCodec("opus")
                    .ForceFormat("webm")
                    .WithCustomArgument("-b:a 128k"))
                .ProcessAsynchronously();
        }
        catch (Exception ex)
        {
            Debug.LogError($"FFmpeg audio streaming error: {ex.Message}");
        }
    }

    IEnumerable<IVideoFrame> GetVideoFrames(CameraStream stream)
    {
        while (isStreaming)
        {
            RenderTexture.active = stream.RenderTexture;
            stream.TempTexture.ReadPixels(new Rect(0, 0, renderWidth, renderHeight), 0, 0);
            stream.TempTexture.Apply();
            RenderTexture.active = null;

            byte[] frameData = stream.TempTexture.GetRawTextureData();
            yield return new RawVideoFrame(frameData, renderWidth, renderHeight, "rgb24");
            yield return null;
        }
    }

    IEnumerable<IAudioSample> GetAudioSamples()
    {
        while (isStreaming)
        {
            lock (audioBuffer)
            {
                int samplesNeeded = 1024 * audioChannels;
                if (audioBuffer.Count >= samplesNeeded)
                {
                    float[] samples = audioBuffer.Take(samplesNeeded).ToArray();
                    audioBuffer.RemoveRange(0, samples.Length);
                    short[] pcmSamples = new short[samples.Length];
                    for (int i = 0; i < samples.Length; i++)
                    {
                        pcmSamples[i] = (short)(samples[i] * 32767f);
                    }
                    yield return new RawAudioSample(pcmSamples, AudioSettings.outputSampleRate, audioChannels, "s16");
                }
            }
            yield return null;
        }
    }

    void Update()
    {
        while (subscriberSocket.TryReceiveFrameString(out string message))
        {
            string[] parts = message.Split(':', 2);
            string command = parts[0];
            string data = parts.Length > 1 ? parts[1] : "";
            string response = ProcessCommand(command, data);
            if (response != null)
            {
                publisherSocket.SendFrame(response);
            }
        }
    }

    string ProcessCommand(string command, string data)
    {
        try
        {
            switch (command)
            {
                case "SET_JOINT_FORCES":
                    var forces = data.Split(',').Select(float.Parse).ToList();
                    List<float> currentForces = new List<float>();
                    articulationBody.GetJointForces(currentForces);
                    for (int i = 0; i < forces.Count && i < currentForces.Count; i++)
                    {
                        currentForces[i] += forces[i];
                    }
                    articulationBody.SetJointForces(currentForces);
                    return "SET_JOINT_FORCES:OK";

                case "GET_DOF_START_INDICES":
                    List<int> indices = new List<int>();
                    articulationBody.GetDofStartIndices(indices);
                    return $"GET_DOF_START_INDICES:{string.Join(",", indices)}";

                case "GET_JOINT_POSITIONS":
                    List<float> positions = new List<float>();
                    articulationBody.GetJointPositions(positions);
                    return $"GET_JOINT_POSITIONS:{string.Join(",", positions)}";

                case "GET_JOINT_VELOCITIES":
                    List<float> velocities = new List<float>();
                    articulationBody.GetJointVelocities(velocities);
                    return $"GET_JOINT_VELOCITIES:{string.Join(",", velocities)}";

                case "GET_JOINT_FORCES":
                    List<float> forces = new List<float>();
                    articulationBody.GetJointForces(forces);
                    return $"GET_JOINT_FORCES:{string.Join(",", forces)}";

                case "GET_JOINT_EXTERNAL_FORCES":
                    List<float> externalForces = new List<float>();
                    int dofCount = articulationBody.GetJointExternalForces(externalForces, Time.fixedDeltaTime);
                    return $"GET_JOINT_EXTERNAL_FORCES:{dofCount},{string.Join(",", externalForces)}";

                default:
                    return $"ERROR:Unknown command:{command}";
            }
        }
        catch (Exception ex)
        {
            return $"ERROR:{ex.Message}";
        }
    }

    void OnDestroy()
    {
        isStreaming = false;
        foreach (var task in ffmpegTasks)
        {
            task.Wait();
        }

        publisherSocket?.Close();
        publisherSocket?.Dispose();
        subscriberSocket?.Close();
        subscriberSocket?.Dispose();
        NetMQConfig.Cleanup(false);

        foreach (var stream in cameraStreams)
        {
            if (stream.RenderTexture) Destroy(stream.RenderTexture);
            if (stream.TempTexture) Destroy(stream.TempTexture);
        }
    }
}